sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsShuffleHeartbeatManager.scala

```diff
@@ -79,30 +115,67 @@ class RapidsShuffleHeartbeatManager extends Logging {
+      registerExecutor(id)
+    } else {
+      val newExecutors = new ArrayBuffer[BlockManagerId]
+      val iter = executors.reverseIterator
+      var done = false
+      while (iter.hasNext && !done) {
+        val entry = iter.next()
+
+        if (entry.registrationOrder >= lastRegistration.lastRegistrationOrderSeen.getValue) {
+          if (entry.id != id) {
+            logDebug(s"Found new executor (to $id): $entry while handling a heartbeat.")
+            newExecutors += entry.id
+          }
+        } else {
+          // We are iterating backwards and have found the last previously inspected executor
+          // since `registrationOrder` monotonically increases and follows insertion order.
+          // We can stop since all peers below this have been sent to the heartbeating peer.
+          done = true
+        }
+      }
+
+      // update this executor's registration with a new heartbeat time, and that last order
+      // from the executors list, indicating the order we should stop at next time
+      lastRegistration.lastHeartbeatMillis.setValue(getCurrentTimeMillis)
+      lastRegistration.lastRegistrationOrderSeen.setValue(registrationOrder)
+
+      // since we updated our heartbeat, update our min-heap
+      leastRecentHeartbeat.priorityUpdated(lastRegistration.id)
+      RapidsExecutorUpdateMsg(newExecutors.toArray)
     }
+  }
 
-    val newExecutors = new ArrayBuffer[BlockManagerId]
-    val iter = executors.zipWithIndex.reverseIterator
-    var done = false
-    while (iter.hasNext && !done) {
-      val (entry, index) = iter.next()
+  // the timeout amount has elapsed, assume this executor is dead
+  private def isStaleHeartbeat(hbMillis: Long, currentTime: Long): Boolean = {
+    (currentTime - hbMillis) > heartbeatTimeoutMillis
+  }
 
-      if (index > lastRegistration.getValue) {
-        if (entry.id != id) {
-          newExecutors += entry.id
+  // must be called holding the object lock
+  private def removeDeadExecutors(currentTime: Long): Unit = {
+    val leastRecentHb = leastRecentHeartbeat.peek() // look at the executor that is lagging most
+    if (leastRecentHb != null &&
+        isStaleHeartbeat(
+          executorRegistrations.get(leastRecentHb).lastHeartbeatMillis.getValue, currentTime)) {
+      // make a new buffer of alive executors and replace the old one
+      val aliveExecutors = new ArrayBuffer[ExecutorRegistration]()
+      executors.foreach { e =>
+        if (isStaleHeartbeat(e.lastHeartbeatMillis.getValue, currentTime)) {
+          logDebug(s"Stale exec, removing $e")
+          executorRegistrations.remove(e.id)
+          leastRecentHeartbeat.remove(e.id)
+        } else {
+          aliveExecutors.append(e)
         }
-      } else {
-        // We are iterating backwards and have found the last index previously inspected
-        // for this peer. We can stop since all peers below this have been sent to this peer.
-        done = true
       }
+      executors.clear()
```

Comment:
Nit: Not seeing the point of clearing an ArrayBuffer we're about to make garbage anyway